api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.trace.propagation;
+
+import static io.opentelemetry.internal.Utils.checkNotNull;
+
+import io.opentelemetry.context.propagation.HttpTextFormat;
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.SpanId;
+import io.opentelemetry.trace.TraceFlags;
+import io.opentelemetry.trace.TraceId;
+import io.opentelemetry.trace.TraceState;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Implementation of the B3 propagation protocol. See <a
+ * href=https://github.com/openzipkin/b3-propagation>openzipkin/b3-propagation</a>.
+ */
+@Immutable
+public class B3Propagator implements HttpTextFormat<SpanContext> {
+  private static final Logger logger = Logger.getLogger(HttpTraceContext.class.getName());
+
+  static final String TRACE_ID_HEADER = "X-B3-TraceId";
+  static final String SPAN_ID_HEADER = "X-B3-SpanId";
+  static final String SAMPLED_HEADER = "X-B3-Sampled";
+  static final String TRUE_INT = "1";
+  static final String FALSE_INT = "0";
+  static final String COMBINED_HEADER = "b3";
+  static final String COMBINED_HEADER_DELIMITER = "-";
+
+  private static final int MAX_TRACE_ID_LENGTH = 2 * TraceId.getSize();
+  private static final int MAX_SPAN_ID_LENGTH = 2 * SpanId.getSize();
+
+  private static final List<String> FIELDS =
+      Collections.unmodifiableList(Arrays.asList(TRACE_ID_HEADER, SPAN_ID_HEADER, SAMPLED_HEADER));
+
+  private final boolean singleHeader;
+
+  /** Creates a new instance of {@link B3Propagator}. Defaults to use multiple headers. */
+  public B3Propagator() {
+    this(false);
+  }
+
+  /**
+   * Creates a new instance of {@link B3Propagator}.
+   *
+   * @param singleHeader whether to use single or multiple headers.
+   */
+  public B3Propagator(Boolean singleHeader) {
```

Comment:
does this need to be a nullable big-b `Boolean`? Can't it just be a primitive boolean?

Reply:
Pushed an update - there is no need for a `Boolean` in this case.